src/facade/dragonfly_connection.cc

```diff
@@ -934,22 +942,31 @@ void Connection::ConnectionFlow(FiberSocketBase* peer) {
   }
 }
 
-void Connection::DispatchCommand(bool has_more, absl::FunctionRef<void()> dispatch_sync,
-                                 absl::FunctionRef<MessageHandle()> dispatch_async) {
+void Connection::DispatchSingle(bool has_more, absl::FunctionRef<void()> dispatch_sync,
+                                absl::FunctionRef<MessageHandle()> dispatch_async) {
   // Avoid sync dispatch if we can interleave with an ongoing async dispatch
   bool can_dispatch_sync = !cc_->async_dispatch;
 
   // Avoid sync dispatch if we already have pending async messages or
   // can potentially receive some (subscriptions > 0)
-  if (dispatch_q_.size() > 0 || cc_->subscriptions > 0)
+  if (dispatch_q_.size() > 0 || cc_->subscriptions > 0) {
     can_dispatch_sync = false;
+    if (stats_->dispatch_queue_bytes >= queue_backpressure_->pipeline_buffer_limit) {
+      DCHECK(queue_backpressure_ == &tl_queue_backpressure_);
+      fb2::NoOpLock noop;
+      queue_backpressure_->pipeline_cnd.wait(noop, [this] {
+        bool res = stats_->dispatch_queue_bytes < queue_backpressure_->pipeline_buffer_limit / 2 ||
+                   cc_->conn_closing;
+        return res;
+      });
+      if (cc_->conn_closing)
+        return;
+    }
```

Comment:
1. why not move it below, above SendAsync? Just for the sake of readability, because the upper part is just determining the dispatch mode
2. Why a jump from limit to limit/2? Can't it make the server freeze fully with a large limit?
3. nit: return "condition" to save on a res variable 😆

Reply:
I am still contemplating about the best behaviour here. I am worried about (lack of ) fairness. 
1. My current code is wrong: when a single connection enters this wait state with `dispatch_queue_bytes > K`, at some point `dispatch_queue_bytes < K`, so other lucky connections will proceed with pipelining without throttling, increase dispatch_queue_bytes, so the stuck connection will starve.
2. Consider the "fixed" scenario: suppose a connection A waits for `dispatch_queue_bytes < K` and it at some fair point it gets activated, because `thread.dispatch_queue_bytes < K`, but then we switch to another connection B that has never waited. It will still be able to steal the turn and continue with pipelining and will increase `thread.dispatch_queue_bytes  > K`. 

Ideally I think we want a mixture of global and local conditions. maybe to wait for the first event that can happen:
a) either the memory will go down or b) `can_dispatch_sync` will evaluate to true because we emptied the LOCAL dispatch_q. 

if b) holds but a) is not we do not have to use pipelining, we can default to sync dispatch.

Reply:
(1) Yep, if the gap is large it might get stuck
(2) That might be a problem only if the client gets stuck exactly at K bytes, which I assume can happen if it get's throttled...